document/encoding/msgpack/encoding.go

```diff
@@ -0,0 +1,306 @@
+package msgpack
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"github.com/genjidb/genji/document"
+	"github.com/vmihailenco/msgpack/v5"
+	"github.com/vmihailenco/msgpack/v5/codes"
+)
+
+// EncodeDocument takes a document and encodes it
+// in MessagePack.
+func EncodeDocument(d document.Document) ([]byte, error) {
+	if ec, ok := d.(EncodedDocument); ok {
+		return ec, nil
+	}
+
+	var buf bytes.Buffer
+	enc := NewEncoder(&buf)
+	defer enc.Close()
+
+	err := enc.EncodeDocument(d)
+	if err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+// DecodeDocument takes a byte slice and returns a lazily decoded document.
+// If buf is malformed, an error will be returned when calling one of the document method.
+func DecodeDocument(buf []byte) document.Document {
+	return EncodedDocument(buf)
+}
+
+// EncodeValue encodes any value to MessagePack.
+func EncodeValue(w io.Writer, v document.Value) error {
+	enc := NewEncoder(w)
+	defer enc.Close()
+
+	return enc.EncodeValue(v)
+}
+
+// An EncodedDocument implements the document.Document
+// interface on top of an encoded representation of a
+// document.
+// It is useful for avoiding decoding the entire document when
+// only a few fields are needed.
+type EncodedDocument []byte
+
+// bytesLen determines the size of the next string in the decoder
+// based on c.
+// It is originally copied from https://github.com/vmihailenco/msgpack/blob/e7759683b74a27e455669b525427cfd9aec0790e/decode_string.go#L10:19
+// then adapted to our needs.
+func bytesLen(c codes.Code, dec *msgpack.Decoder) (int, error) {
+	if c == codes.Nil {
+		return -1, nil
+	}
+
+	if codes.IsFixedString(c) {
+		return int(c & codes.FixedStrMask), nil
+	}
+
+	switch c {
+	case codes.Str8, codes.Bin8:
+		var b [1]byte
+		err := dec.ReadFull(b[:])
+		return int(b[0]), err
+	case codes.Str16, codes.Bin16:
+		var b [2]byte
+		err := dec.ReadFull(b[:])
+		return int((uint16(b[0]) << 8) | uint16(b[1])), err
+	case codes.Str32, codes.Bin32:
+		var b [4]byte
+		err := dec.ReadFull(b[:])
+		n := (uint32(b[0]) << 24) |
+			(uint32(b[1]) << 16) |
+			(uint32(b[2]) << 8) |
+			uint32(b[3])
+		return int(n), err
+	}
+
+	return 0, fmt.Errorf("msgpack: invalid code=%x decoding bytes length", c)
+}
+
+// GetByField decodes the selected field from the buffer.
+func (e EncodedDocument) GetByField(field string) (v document.Value, err error) {
+	dec := NewDecoder(bytes.NewReader(e))
+	defer dec.Close()
+
+	l, err := dec.dec.DecodeMapLen()
+	if err != nil {
+		return
+	}
+
+	bf := []byte(field)
+
+	buf := make([]byte, 32)
```

Comment:
Limiting the *number* of allocations is more important than the size of buffers (reading how `append` is implemented is a good example of this rule), because the less allocations, the less work the garbage collector has to do.

If we have this kind of document, where keys are only slightly bigger from one another:

```js
{
  "a": 1,
  ....
  "aa": 2,
  ....
  "aaaaaa": 3,
}
```

We might end up creating 3 or 4 different slices, which means 3 or 4 objects tracked by the GC.

With a larger buffer, we only need to allocate it once most of the time, which means potentially less objects tracked by the GC.